### PR TITLE
add JSON_MERGE_PRESERVE() function

### PR DESCRIPTION
**Pull request by jennifersp**

### CODE REVIEW

File: sql/json.go [hunk 1]
```diff
@@ -107,3 +107,71 @@ func (t jsonType) Zero() interface{} {
 	// JSON Null
 	return nil
 }
+
+// Copy implements deep copy
+func Copy(v interface{}) interface{} {
```
Comment (andrew-wm-arthur) [inline sql/json.go:112]:
```suggestion
func DeepCopyJson(v interface{}) interface{} {
```

File: sql/json.go [hunk 2]
```diff
@@ -107,3 +107,71 @@ func (t jsonType) Zero() interface{} {
...
+		return newUInt8
+	case uint16:
+		newUInt16 := v.(uint16)
+		return newUInt16
+	case uint32:
+		newUInt32 := v.(uint32)
+		return newUInt32
+	case uint64:
+		newUInt64 := v.(uint64)
+		return newUInt64
```
Comment (andrew-wm-arthur) [inline sql/json.go:173]: these cases handling scalars can be collapsed into one:
```suggestion
	case bool, string, float64, float32, 
		int, int8, int16, int32, int64, 
		uint, uint8, uint16, uint32, uint64:
		return v
```

**Review by andrew-wm-arthur — COMMENTED** [review]
Generally looks good. Be sure to add some enginetests

File: sql/expression/function/json_merge_preserve.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+// Description implements sql.FunctionExpression
+func (j *JSONMergePreserve) Description() string {
+	return "merges JSON documents, preserving duplicate keys."
+}
+
+// IsUnsupported implements sql.UnsupportedFunctionStub
+func (j JSONMergePreserve) IsUnsupported() bool {
+	return true
```
Comment (VinaiRachakonda) [inline sql/expression/function/json_merge_preserve.go:73]: This function looks supported to me :)

File: sql/expression/function/json_merge_preserve.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+func (j *JSONMergePreserve) Resolved() bool {
```
Comment (VinaiRachakonda) [inline sql/expression/function/json_merge_preserve.go:76]: Missing comment

File: sql/expression/function/json_merge_preserve.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	for _, d := range j.JSONDocs {
+		if !d.Resolved() {
+			return false
+		}
+	}
+	return true
+}
+
+func (j *JSONMergePreserve) String() string {
```
Comment (VinaiRachakonda) [inline sql/expression/function/json_merge_preserve.go:85]: Missing comment

File: sql/expression/function/json_merge_preserve.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	var parts = make([]string, len(children))
+
+	for i, c := range children {
+		parts[i] = c.String()
+	}
+
+	return fmt.Sprintf("JSON_MERGE_PRESERVE(%s)", strings.Join(parts, ", "))
+}
+
+func (j *JSONMergePreserve) Type() sql.Type {
```
Comment (VinaiRachakonda) [inline sql/expression/function/json_merge_preserve.go:96]: Be sure to add the top level comment for each function

File: sql/expression/function/json_merge_preserve_test.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+	f4, err := NewJSONMergePreserve(
+		expression.NewGetField(0, sql.LongText, "arg1", false),
+		expression.NewGetField(1, sql.LongText, "arg2", false),
+		expression.NewGetField(2, sql.LongText, "arg3", false),
+		expression.NewGetField(3, sql.LongText, "arg4", false),
+	)
+	require.NoError(t, err)
+
+	jsonArray1 := []interface{}{1, 2}
```
Comment (VinaiRachakonda) [inline sql/expression/function/json_merge_preserve_test.go:48]: I'd like to see some more tests that use `null` JSON values.

File: sql/json.go [hunk 8]
```diff
@@ -107,3 +107,33 @@ func (t jsonType) Zero() interface{} {
 	// JSON Null
 	return nil
 }
+
+// DeepCopyJson implements deep copy of JSON document
+func DeepCopyJson(v interface{}) interface{} {
```
Comment (VinaiRachakonda) [inline sql/json.go:112]: Hell yea!

**Review by VinaiRachakonda — APPROVED** [review]
LGTM, double check mysql parity and add some more tests that work with nulls
